File: go/examples/manage-indexes/drop-index.go
//	:replace-start: {
//	   "terms": {
//	      "manage_indexes": "main",
//	      "ExampleDropIndex": "main"
//	   }
//	}
//
// :snippet-start: examples
package manage_indexes

import (
	"context"
	"fmt"
	"log"
	"os"   // :remove:
	"time" // :remove:

	"github.com/joho/godotenv" // :remove:

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

func ExampleDropIndex() {
	ctx := context.Background()
	// :remove-start:
	if err := godotenv.Load("../../.env"); err != nil {
		log.Println("no .env file found")
	}
	// Connect to your Atlas cluster
	uri := os.Getenv("ATLAS_CONNECTION_STRING")
	if uri == "" {
		log.Fatal("set your 'ATLAS_CONNECTION_STRING' environment variable.")
	}
	// :remove-end:
	// Replace the placeholder with your Atlas connection string
	// :uncomment-start:
	//const uri = "<connection-string>"
	// :uncomment-end:

	// Connect to your Atlas cluster
	clientOptions := options.Client().ApplyURI(uri)
	client, err := mongo.Connect(ctx, clientOptions)
	if err != nil {
		log.Fatalf("failed to connect to the server: %v", err)
	}
	defer func() { _ = client.Disconnect(ctx) }()

	// Set the namespace
	coll := client.Database("sample_mflix").Collection("embedded_movies")
	indexName := "vector_index"

	err = coll.SearchIndexes().DropOne(ctx, indexName)
	if err != nil {
		log.Fatalf("failed to delete the index: %v", err)
	}

	fmt.Println("Successfully deleted the Vector Search index")
	// :remove-start:
	fmt.Println("Polling to confirm successful index deletion.")
	fmt.Println("NOTE: This may take up to a minute.")
	searchIndexes := coll.SearchIndexes()
	indexNotYetDeleted := true
	loopNumber := 0
	for indexNotYetDeleted {
		cursor, err := searchIndexes.List(ctx, options.SearchIndexes().SetName(indexName))
		if err != nil {
			fmt.Errorf("failed to list search indexes: %w", err)
		}

		if !cursor.Next(ctx) {
			break
		}

		name := cursor.Current.Lookup("name").StringValue()
		// If dropping the index takes more than a minute, which is 12 loops
		// with a 5 second sleep, something has gone wrong and we should
		// abandon all hope
		if name == indexName && loopNumber < 12 {
			time.Sleep(5 * time.Second)
			loopNumber += 1
		} else if name == indexName && loopNumber == 12 {
			log.Fatalf("Attempted to drop the index for a minute but it's still there. Something went wrong.")
			indexNotYetDeleted = false
		} else {
			indexNotYetDeleted = false
		}
	}
	fmt.Println("Index named " + indexName + " was deleted.")
	// :remove-end:
}

// :snippet-end:
// :replace-end:

File: java/src/main/java/indexes/DropIndex.java
//	:replace-start: {
//	  "terms": {
//	    "System.getenv(\"ATLAS_CONNECTION_STRING\")": "<connectionString>"
//	  }
//	}
package indexes;
// :snippet-start: example
import com.mongodb.client.ListSearchIndexesIterable;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoCursor;
import com.mongodb.client.MongoDatabase;
import org.bson.Document;

public class DropIndex {
    public static void main(String[] args) {
        // Replace the placeholder with your Atlas connection string
        String uri = System.getenv("ATLAS_CONNECTION_STRING");

        // Connect to your Atlas cluster
        try (MongoClient mongoClient = MongoClients.create(uri)) {

            // Set the namespace
            MongoDatabase database = mongoClient.getDatabase("sample_mflix");
            MongoCollection<Document> collection = database.getCollection("embedded_movies");

            // Specify the index to delete
            String indexName = "vector_index";

            try {
                collection.dropSearchIndex(indexName);
            } catch (Exception e) {
                throw new RuntimeException("Error deleting index: " + e);
            }
            // :remove-start:
            // Wait for the drop index operation to complete
            System.out.println("Polling to confirm the index has successfully been deleted.");

            ListSearchIndexesIterable<Document> searchIndexes = collection.listSearchIndexes();
            boolean isDeleted = false;
            Document doc = null;
            while (!isDeleted) {
                boolean indexMatchingNameExists = false;
                ListSearchIndexesIterable<Document> innerSearchIndexes = collection.listSearchIndexes();
                try (MongoCursor<Document> cursor = innerSearchIndexes.iterator()) {
                    if (!cursor.hasNext()) {
                        break;
                    }
                    Document current = cursor.next();
                    String name = current.getString("name");
                    if (name.equals(indexName)) {
                        indexMatchingNameExists = true;
                    }
                    if (indexMatchingNameExists) {
                        Thread.sleep(500);
                    } else {
                        isDeleted = true;
                    }
                } catch (Exception e) {
                    throw new RuntimeException("Failed to list search indexes: " + e);
                }
            }
            System.out.println(indexName + " index has successfully been deleted.");
            // :remove-end:
        } catch (Exception e) {
            throw new RuntimeException("Error connecting to MongoDB: " + e);
        }
    }
}
// :snippet-end:
// :replace-end:
